scipy/sparse/csgraph/_disjoint_set.py

```diff
@@ -0,0 +1,177 @@
+    1
+    >>> disjoint_set['b']
+    3
+
+    Test connectivity:
+
+    >>> disjoint_set.connected(1, 2)
+    True
+    >>> disjoint_set.connected(1, 'b')
+    False
+
+    List elements in disjoint set:
+
+    >>> list(disjoint_set)
+    [1, 2, 3, 'a', 'b']
+    """
+    def __init__(self):
+        self.n_nodes = 0
+        self.n_components = 0
+        self._sizes = {}
+        self._parents = {}
+        self._indices = collections.OrderedDict()
+
+    def __iter__(self):
+        """Returns an iterator of the nodes in the disjoint set.
+
+        Nodes in the disjoint set consist of all inputs to the `__getitem__`,
+        `merge`, and `connected` methods. Elements are ordered by insertion
+        order.
+        """
+        return iter(self._indices)
+
+    def __getitem__(self, x):
+        """Find the root node of `x`.
+
+        The node is added to the disjoint set if not already present, in which
+        case the node is its own root.
+
+        Parameters
+        ----------
+        x : hashable object
+            Input node.
+
+        Returns
+        -------
+        root : hashable object
+            Root node of `x`.
+        """
+        if x not in self._parents:
+            # add node
+            self._sizes[x] = 1
+            self._parents[x] = x
+            self._indices[x] = len(self._indices)
+            self.n_nodes += 1
+            self.n_components += 1
+            return x
+
+        # find by "path halving"
+        parents = self._parents
+        while self._indices[x] != self._indices[parents[x]]:
+            parents[x] = parents[parents[x]]
+            x = parents[x]
+        return x
+
+    def merge(self, x, y):
+        """Merge the subsets of `x` and `y`.
+
+        The smaller subset (the child) is merged into the larger subset (the
+        parent). If the subsets are of equal size, the root node which was
+        first inserted into the disjoint set is selected as the parent.
+
+        Nodes which are not already present are added to the disjoint set. If        
+        neither node is present, `x` is inserted into the disjoint set before
+        `y`.
+
+        Parameters
+        ----------
+        x, y : hashable object
+            Nodes to merge.
+
+        Returns
+        -------
+        merged : bool
+            True if `x` and `y` were in disjoint sets, False otherwise.
+        """
+        x = self[x]
+        y = self[y]
+        if self._indices[x] == self._indices[y]:
+            return False
+
+        sizes = self._sizes
+        if (sizes[x], self._indices[y]) < (sizes[y], self._indices[x]):
+            x, y = y, x
+        self._parents[y] = x
+        self._sizes[x] += self._sizes[y]
+        self.n_components -= 1
+        return True
+
+    def connected(self, x, y):
+        """Test whether `x` and `y` are in the same component/set.
+
+        If neither `x` nor `y` has been previously seen, `x` is
+        inserted into the disjoint set before `y`.
```

Comment:
It's weird to me that this modifies the state which can break things like this.

```
>>> import disjoint
>>> dis = disjoint.DisjointSet()
>>> dis.merge(1, 10)
True
>>> dis.merge(2, 20)
True

>>> for a in dis:
...     if dis.connected(a, 30): print("True")
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: OrderedDict mutated during iteration
```

Reply:
Nice catch. I have changed the behaviour of `__iter__` so that it only iterates over nodes which are present at the time of instantiation. A unit test has been added too.